### PR TITLE
For subworkflows, no outputs means no outputs.

### PR DESCRIPTION
**Pull request by cjllanwarne**

### CODE REVIEW

File: wdl/src/main/scala/wdl/WdlWorkflow.scala [hunk 1]
```diff
@@ -155,15 +155,21 @@ case class WdlWorkflow(unqualifiedName: String,
     leftOutputs.find(_.unqualifiedName == name)
   }
 
-  lazy val hasEmptyOutputSection = workflowOutputWildcards.isEmpty && children.collect({ case o: WorkflowOutput => o }).isEmpty
+  lazy val noWorkflowOutputs = children.collect { case o: WorkflowOutput => o }.isEmpty
+  lazy val noOutputWildcards = workflowOutputWildcards.isEmpty
+
+  lazy val hasEmptyOutputSection =  noWorkflowOutputs && noOutputWildcards
+
+  lazy val isTopLevelWorkflow: Boolean =
```
Comment (geoffjentry) [inline wdl/src/main/scala/wdl/WdlWorkflow.scala]: In general I've been trying to be good about not chiming in on such quibbles but can we not have this on two lines at least? :)

File: wdl/src/main/scala/wdl/WdlWorkflow.scala [hunk 2]
```diff
@@ -155,15 +155,21 @@ case class WdlWorkflow(unqualifiedName: String,
...
 
   /**
-   * All outputs for this workflow and their associated types
+   * All outputs for this workflow and their associated types. Only applies to top-level workflows.
    *
    * @return a Map[FullyQualifiedName, WomType] representing the union
    *         of all outputs from all `call`s within this workflow
    */
-  lazy val expandedWildcardOutputs: Seq[WorkflowOutput] = {
+  lazy val expandedWildcardOutputs: Seq[WorkflowOutput] = if(!ancestry.exists(_.isInstanceOf[WdlWorkflow])) Seq.empty else {
```
Comment (geoffjentry) [inline wdl/src/main/scala/wdl/WdlWorkflow.scala]: likewise here, the if and else sides are so imbalanced that it'd be a lot easier to follow if the else side was pulled into its own function or something.

**Comment by geoffjentry** [comment]
:+1: modulo some old school jeff nit pickery

File: wdl/src/main/scala/wdl/WdlWorkflow.scala (hunk 2 above)
Comment (cjllanwarne) [reply wdl/src/main/scala/wdl/WdlWorkflow.scala]: Also: this should be using `workflow.isTopLevelWorkflow`, not this weird not-quite-right hack

File: wdl/src/main/scala/wdl/WdlWorkflow.scala (hunk 1 above)
Comment (cjllanwarne) [reply wdl/src/main/scala/wdl/WdlWorkflow.scala]: Oh yeah, this was to make it easier to target the debugger, should be reverted

File: wdl/src/main/scala/wdl/WdlGraphNode.scala [hunk 3]
```diff
@@ -146,7 +146,11 @@ object WdlGraphNode {
       _.inputPorts.map(_.upstream.graphNode).filterByType[OuterGraphInputNode]: Set[OuterGraphInputNode]
     }
 
-    def withDefaultOutputs(g: Graph): Graph = if (g.nodes.exists(_.isInstanceOf[GraphOutputNode])) { g } else {
+    // Default outputs are added if we're:
+    // - A scatter block
+    // - An if block
+    // - (NB: top level workflows are already given wildcard outputs in the WDL Workflow building phase)
+    def withDefaultOutputs(g: Graph): Graph = if (from.isInstanceOf[If] || from.isInstanceOf[Scatter]) {
```
Comment (mcovarr) [inline wdl/src/main/scala/wdl/WdlGraphNode.scala]: uh can that be done in a `collect` or something that isn't `isInstanceOf`ing?

File: wdl/src/main/scala/wdl/WdlWorkflow.scala [hunk 4]
```diff
@@ -155,15 +155,22 @@ case class WdlWorkflow(unqualifiedName: String,
...
 
   /**
-   * All outputs for this workflow and their associated types
+   * All outputs for this workflow and their associated types. Only applies to top-level workflows.
    *
    * @return a Map[FullyQualifiedName, WomType] representing the union
    *         of all outputs from all `call`s within this workflow
    */
-  lazy val expandedWildcardOutputs: Seq[WorkflowOutput] = {
+  lazy val expandedWildcardOutputs: Seq[WorkflowOutput] = if(isTopLevelWorkflow) calculateExpandedWildcardOutputs else Seq.empty
```
Comment (mcovarr) [inline wdl/src/main/scala/wdl/WdlWorkflow.scala]: space

File: wdl/src/main/scala/wdl/WdlWorkflow.scala [hunk 5]
```diff
@@ -155,15 +155,22 @@ case class WdlWorkflow(unqualifiedName: String,
     leftOutputs.find(_.unqualifiedName == name)
   }
 
-  lazy val hasEmptyOutputSection = workflowOutputWildcards.isEmpty && children.collect({ case o: WorkflowOutput => o }).isEmpty
+  lazy val noWorkflowOutputs = children.collect { case o: WorkflowOutput => o }.isEmpty
+  lazy val noOutputWildcards = workflowOutputWildcards.isEmpty
+
+  lazy val hasEmptyOutputSection =  noWorkflowOutputs && noOutputWildcards
```
Comment (mcovarr) [inline wdl/src/main/scala/wdl/WdlWorkflow.scala]: less space

**Comment by cjllanwarne** [comment]
NB the failures in `travis/pr` `centaurJes` were caused by the `customLabels` change merging into develop before I rebased. 

I didn't want to wait 30 minutes to rerun the tests, since they were already passing on the other backends, and in `travis/push`...

**Pull request merged**